File: python-algorithm/leetcode/problem_307.py
"""307. Range Sum Query - Mutable
https://leetcode.com/problems/range-sum-query-mutable/
"""
from typing import List


class NumArray:

    def __init__(self, nums: List[int]):
        n = len(nums)
        size = int(n ** 0.5)
        sums = [0] * ((n + size - 1) // size)
        for i, num in enumerate(nums):
            sums[i // size] += num
        self.nums = nums
        self.sums = sums
        self.size = size

    def update(self, index: int, val: int) -> None:
        self.sums[index // self.size] += val - self.nums[index]
        self.nums[index] = val

    def sumRange(self, left: int, right: int) -> int:
        block1, block2 = left // self.size, right // self.size
        if block1 == block2:
            return sum(self.nums[left:right + 1])
        return sum(self.nums[left:(block1 + 1) * self.size]) + \
               sum(self.sums[block1 + 1: block2]) + \
               sum(self.nums[block2 * self.size: right + 1])

File: java-algo/src/main/java/com/leetcode/Problem307.java
package com.leetcode;

/**
 * 307. Range Sum Query - Mutable
 * https://leetcode.com/problems/range-sum-query-mutable/
 */
public class Problem307 {
    private static class NumArray {
        int[] tree;
        int n;
        int[] nums;

        int lowbit(int x) {
            return x & -x;
        }

        int query(int x) {
            int ret = 0;
            for (int i = x; i > 0; i -= lowbit(i)) {
                ret += tree[i];
            }
            return ret;
        }

        void add(int x, int v) {
            for (int i = x; i <= n; i += lowbit(i)) {
                tree[i] += v;
            }
        }
        public NumArray(int[] nums) {
            this.nums = nums;
            this.n = nums.length;
            tree = new int[n + 1];
            for (int i = 0; i < n; i++) {
                add(i + 1, nums[i]);
            }
        }

        public void update(int index, int val) {
            add(index + 1, val - nums[index]);
            nums[index] = val;
        }

        public int sumRange(int left, int right) {
            return query(right + 1) - query(left);
        }
    }
}
